File: talk.lua
local p = {}

local dataCache
local templateName = '简单对话_TEST'

local chapterPages = {
    ['序章/剧情'] = 'prologue',
    ['万圣夜/剧情'] = 'halloween',
    ['梦境世界/剧情'] = 'dream',
    ['辉针城/剧情'] = 'castle',
    ['梦殿大祀庙/剧情'] = 'mausoleum',
    ['地灵殿/剧情'] = 'palace',
    ['圣辇船/剧情'] = 'ship',
}

local function loadData()
    if dataCache then return dataCache end
    local ok, data = pcall(mw.loadJsonData, 'Talk.json')
    if not ok or type(data) ~= 'table' then
        local title = mw.title.new('Talk.json')
        if not title or not title.exists then return nil end
        ok, data = pcall(mw.text.jsonDecode, title:getContent() or '')
        if not ok or type(data) ~= 'table' then return nil end
    end
    dataCache = data.chapters or data
    return dataCache
end

local function trim(value)
    return mw.text.trim(tostring(value or ''))
end

local function getArg(frame)
    local parent = frame:getParent()
    if trim(frame.args[1]) ~= '' then return trim(frame.args[1]) end
    if parent then return trim(parent.args[1]) end
    return ''
end

local function escapeTemplateText(value)
    local text = trim(value):gsub('|', '{{!}}')
    text = text:gsub('<br%s*/?>', '<br>')
    text = text:gsub('<color=([^>]+)>(.-)</color>', '{{color|%1|%2}}')
    text = text:gsub('<b>(.-)</b>', "'''%1'''")
    text = text:gsub('<i>(.-)</i>', "''%1''")
    text = text:gsub('<u>(.-)</u>', '<u>%1</u>')
    text = text:gsub('<size=[^>]+>(.-)</size>', '%1')
    return text
end

local function dialogue(style, speaker, text)
    return string.format(
        '{{%s|style=%s|%s|%s}}',
        templateName,
        tostring(style or 1),
        escapeTemplateText(speaker),
        escapeTemplateText(text)
    )
end

local function event(text)
    return string.format('{{%s|style=7|%s}}', templateName, escapeTemplateText(text))
end

local function bgmEvent(text)
    return string.format('{{%s|style=6|%s}}', templateName, escapeTemplateText(text))
end

local function sectionTag(kind, name)
    local label = trim(name)
    label = label:gsub('&', '&amp;')
    label = label:gsub('"', '&quot;')
    label = label:gsub('<', '&lt;')
    label = label:gsub('>', '&gt;')
    return string.format('<section %s="%s" />', kind, label)
end

local function characterEvent(characters)
    if type(characters) ~= 'table' or #characters == 0 then return nil end
    local names = {}
    for _, character in ipairs(characters) do
        local name = trim(character.name or character.id)
        if name ~= '' and character.side ~= 'self' then
            table.insert(names, '[[' .. name .. ']]')
        end
    end
    if #names == 0 then return nil end
    local joined
    if #names == 1 then
        joined = names[1]
    else
        joined = table.concat(names, '、', 1, #names - 1) .. ' 与 ' .. names[#names]
    end
    return event(joined .. ' 登场')
end

local function renderSection(section)
    local result = {}
    if trim(section.name) ~= '' then table.insert(result, event(section.name)) end
    local appearance = characterEvent(section.characters)
    if appearance then table.insert(result, appearance) end
    for _, sentence in ipairs(section.sentences or {}) do
        if trim(sentence.description) ~= '' then
            table.insert(result, event(sentence.description))
        end
        table.insert(result, dialogue(sentence.style, sentence.name or sentence.speaker, sentence.text))
    end
    return table.concat(result, '\n')
end

local function renderGroup(group, showTitle)
    local name = trim(group.name or group.id)
    local result = {}
    if showTitle then
        table.insert(result, '== ' .. name .. ' ==')
    end
    table.insert(result, sectionTag('begin', name))
    if trim(group.music) ~= '' then
        table.insert(result, bgmEvent('BGM：' .. trim(group.music) .. ''))
    end
    for _, section in ipairs(group.sections or {}) do
        local text = renderSection(section)
        if text ~= '' then table.insert(result, text) end
    end
    table.insert(result, sectionTag('end', name))
    return table.concat(result, '\n')
end

local function renderChapter(chapter)
    local result = {}
    for _, group in ipairs(chapter.groups or {}) do
        table.insert(result, renderGroup(group, true))
    end
    return table.concat(result, '\n\n')
end

function p.getTalk(frame)
    local data = loadData()
    if not data then return '错误：无法加载 [[Talk.json]]' end

    local argument = getArg(frame)
    local chapterKey = chapterPages[argument]

    if chapterKey then
        for _, chapter in ipairs(data) do
            if chapter.key == chapterKey then
                return frame:preprocess(renderChapter(chapter))
            end
        end
        return '错误：[[Talk.json]] 中缺少「' .. argument .. '」的数据。'
    end

    local matchedGroup
    for _, chapter in ipairs(data) do
        for _, group in ipairs(chapter.groups or {}) do
            if trim(group.name) == argument then
                if matchedGroup then
                    return '错误：存在多个名为「' .. argument .. '」的小节。'
                end
                matchedGroup = group
            end
        end
    end

    if matchedGroup then
        return frame:preprocess(renderGroup(matchedGroup, false))
    end
    return '错误：参数「' .. argument .. '」不是剧情页面或小节名称。'
end

return p

File: talk.py
import json
import os
import re
import xml.etree.ElementTree as ET
from collections import Counter, defaultdict


BASE_DIR = os.path.dirname(os.path.abspath(__file__))
METAS_DIR = os.path.join(BASE_DIR, "metas")
TALKS_DIR = os.path.join(METAS_DIR, "talks")
CHARACTERS_PATH = os.path.join(METAS_DIR, "talkcharacters.xml")
MUSICS_PATH = os.path.join(METAS_DIR, "musics.xml")
NS = "mvz2:"
WIKI_JSON_TITLE = "Talk.json"

CHAPTERS = {
    "prologue": "序章",
    "halloween": "万圣夜",
    "dream": "梦境世界",
    "castle": "辉针城",
    "mausoleum": "梦殿大祀庙",
    "ship": "圣辇船",
    "palace": "地灵殿",
}


def short(value):
    if isinstance(value, str) and value.startswith(NS):
        return value[len(NS):]
    return value


def compact_dict(values):
    return {
        key: value
        for key, value in values.items()
        if value not in (None, "", {}, [])
    }


def read_xml(path):
    with open(path, encoding="utf-8-sig") as source:
        return ET.fromstring(source.read())


def load_character_names():
    root = read_xml(CHARACTERS_PATH)
    names = {}
    for character in root.findall("character"):
        character_id = short(character.get("id", "").strip())
        name = character.get("name", "").strip()
        if character_id and name:
            names[character_id] = name
    return names


def load_music_names():
    root = read_xml(MUSICS_PATH)
    names = {}
    for music in root.findall("music"):
        name = music.get("name", "").strip()
        if not name:
            continue

        music_id = short(music.get("id", "").strip())
        if music_id:
            names[music_id] = name

        track = music.find("track")
        if track is not None:
            for track_id in track.attrib.values():
                track_id = short(track_id.strip())
                if track_id:
                    names[track_id] = name
    return names


def clean_text(value):
    if value is None:
        return ""
    return re.sub(r"\s+", " ", value).strip()


def sentence_text(sentence):
    text = sentence.find("text")
    return clean_text("".join(text.itertext())) if text is not None else ""


def script_sides(root):
    sides = defaultdict(Counter)
    for character in root.findall(".//characters/character"):
        character_id = short(character.get("id", ""))
        side = character.get("side")
        if character_id and side:
            sides[character_id][side] += 1

    pattern = re.compile(r"^character\s+create\s+(\S+)\s+(left|right|self)(?:\s|$)")
    for script in root.findall(".//script"):
        match = pattern.match(clean_text(script.text))
        if match:
            sides[short(match.group(1))][match.group(2)] += 1
    return sides


def preferred_sides(roots):
    totals = defaultdict(Counter)
    for root in roots:
        for character_id, counts in script_sides(root).items():
            totals[character_id].update(counts)
    return {
        character_id: counts.most_common(1)[0][0]
        for character_id, counts in totals.items()
    }


def style_for(side):
    return {"left": 1, "right": 2, "self": 3}.get(side, 1)


def parse_section(section, names, defaults):
    local_sides = {}
    characters = []
    for character in section.findall("./characters/character"):
        character_id = short(character.get("id", ""))
        side = character.get("side") or defaults.get(character_id, "left")
        local_sides[character_id] = side
        characters.append(compact_dict({
            "id": character_id,
            "name": names.get(character_id, character_id),
            "side": side,
            "variant": short(character.get("variant")),
        }))

    sentences = []
    for sentence in section.findall("./sentences/sentence"):
        speaker_id = short(sentence.get("speaker", ""))
        side = local_sides.get(speaker_id) or defaults.get(speaker_id, "left")
        sentences.append(compact_dict({
            "speaker": speaker_id,
            "name": names.get(speaker_id, speaker_id or "？？？"),
            "side": side,
            "style": style_for(side),
            "text": sentence_text(sentence),
            "description": clean_text(sentence.get("description")),
            "variant": short(sentence.get("variant")),
        }))

    section_name = section.findtext("text")
    return compact_dict({
        "name": clean_text(section_name),
        "characters": characters,
        "sentences": sentences,
    })


def parse_chapter(key, root, names, music_names, defaults):
    groups = []
    for group in root.findall("group"):
        archive = group.find("archive")
        group_name = archive.get("name", "") if archive is not None else group.get("id", "")
        music_id = short(archive.get("music")) if archive is not None else None
        groups.append(compact_dict({
            "id": group.get("id"),
            "name": group_name,
            "tags": [short(tag.strip()) for tag in group.get("tags", "").split(";") if tag.strip()],
            "music": music_names.get(music_id, music_id),
            "sections": [
                parse_section(section, names, defaults)
                for section in group.findall("section")
            ],
        }))

    try:
        order = int(root.get("order", "0"))
    except ValueError:
        order = root.get("order", "0")
    return {
        "key": key,
        "name": CHAPTERS[key],
        "file": key + ".xml",
        "order": order,
        "groups": groups,
    }


def convert():
    names = load_character_names()
    music_names = load_music_names()
    roots = {}
    for key in CHAPTERS:
        path = os.path.join(TALKS_DIR, key + ".xml")
        if not os.path.exists(path):
            raise FileNotFoundError("缺少剧情 XML：" + path)
        roots[key] = read_xml(path)

    defaults = preferred_sides(roots.values())
    chapters = [
        parse_chapter(key, root, names, music_names, defaults)
        for key, root in roots.items()
    ]
    chapters.sort(key=lambda chapter: chapter["order"])
    return json.dumps(
        {"chapters": chapters},
        ensure_ascii=False,
        separators=(",", ":"),
    )


if __name__ == "__main__":
    print(convert())